windows/RNScreens/Screen.cpp

```diff
@@ -0,0 +1,117 @@
+#include "pch.h"
+#include "NativeModules.h"
+#include "JSValueXaml.h"
+#include "Screen.h"
+
+namespace winrt {
+    using namespace Microsoft::ReactNative;
+    using namespace Windows::Foundation;
+    using namespace Windows::Foundation::Collections;
+    using namespace Windows::UI;
+    using namespace Windows::UI::Xaml;
+    using namespace Windows::UI::Xaml::Controls;
+}
+
+namespace winrt::RNScreens::implementation {
+    Screen::Screen(winrt::Microsoft::ReactNative::IReactContext reactContext)
+        : m_reactContext(reactContext) {
+        onLoadingRevoker = Loading({ this, &Screen::onLoading });
+        onLoadedRevoker = Loaded({ this, &Screen::onLoaded });
+        onUnloadedRevoker = Unloaded({ this, &Screen::onUnloaded });
+    }
+
+    Screen::~Screen() {
+        Loading(onLoadingRevoker);
+        Loaded(onLoadedRevoker);
+        Unloaded(onUnloadedRevoker);
+    }
+
+    void Screen::addView(winrt::Windows::UI::Xaml::UIElement element) {
+        Children().Append(element);
+    }
+
+    void Screen::removeAllChildren() {
+        Children().Clear();
+    }
+
+    void Screen::removeChildAt(int64_t index) {
+        Children().RemoveAt(static_cast<uint32_t>(index));
+    }
+
+    void Screen::replaceChild(winrt::Windows::UI::Xaml::UIElement oldChild,
+        winrt::Windows::UI::Xaml::UIElement newChild) {
+        uint32_t index;
+        if (!Children().IndexOf(oldChild, index))
+            return;
+
+        Children().SetAt(index, newChild);
+    }
+
+    void Screen::onLoading(winrt::Windows::UI::Xaml::FrameworkElement const& sender,
+        winrt::Windows::Foundation::IInspectable const&) {
+        auto screen = sender.try_as<Screen>();
+        if (!screen)
+            return;
+
+        screen->dispatchOnWillAppear();
+    }
+
+    void Screen::onLoaded(winrt::Windows::Foundation::IInspectable const& sender,
+        winrt::Windows::UI::Xaml::RoutedEventArgs const&) {
+        auto screen = sender.try_as<Screen>();
+        if (!screen)
+            return;
+
+        screen->dispatchOnAppear();
+    }
+
+    void Screen::onUnloaded(winrt::Windows::Foundation::IInspectable const& sender,
+        winrt::Windows::UI::Xaml::RoutedEventArgs const&) {
+        auto screen = sender.try_as<Screen>();
+        if (!screen)
+            return;
+
+        screen->dispatchOnWillDisappear();

```

Comment:
To note that the Android implementation does the same thing: https://github.com/software-mansion/react-native-screens/blob/b5777665392badbc3f175535f3769bbd3653f2e5/android/src/main/java/com/swmansion/rnscreens/ScreenStackFragment.java#L157

Since I didn't find anything more appropriate on Windows, I just did the same thing as the Android implementation.

Reply:
hmm this is only in the case of no screen transition animation: https://github.com/software-mansion/react-native-screens/blob/b5777665392badbc3f175535f3769bbd3653f2e5/android/src/main/java/com/swmansion/rnscreens/ScreenStackFragment.java#L135. Here it is dispatched at the beginning of the animation: https://github.com/software-mansion/react-native-screens/blob/b5777665392badbc3f175535f3769bbd3653f2e5/android/src/main/java/com/swmansion/rnscreens/ScreenFragment.java#L195.

Reply:
I see, I found it a bit hard to navigate through some of that code, I just tried to follow it as best as I could with my limited understanding.

Reply:
Yeah, we use some hacks due to buggy behaviors of lifecycle methods on Android. The `onWillAppear`/`onWillDisappear` lifecycle methods should be dispatched at the beginning of the transition between screens, and `onAppear`/`onDisappear` at the end of transition.